### PR TITLE
Extend libbpf-tools for Inspektor Gadget

### PR DESCRIPTION
**Pull request by mauriciovasquezbernal**
This PR is a POC to make it possible to libbpf-based tools with Inspektor Gadget

The changes in this PR are:
- Extend execsnoop to be able to filter by container and print container info
- Package libbpf-tool in the BCC container image 

TODO
- [x] It's not clear how to handle the different docker images. Currently 3 versions are pushed: core, standard and gadget that combines the first two. 
- [x] Extend other tools used in Inspektor Gadget
- [ ] Implement --json support?

### CODE REVIEW

**Comment by mauriciovasquezbernal** [comment]
@alban in this PR three different container image tags are created: one for the standard tools, one for libbpf-based tools and another one that combines those two. Do you have any feedback on this one?

**Comment by alban** [comment]
> @alban in this PR three different container image tags are created: one for the standard tools, one for libbpf-based tools and another one that combines those two. Do you have any feedback on it?

I am not sure to understand why we need to build those 3 flavours. Couldn't we just build the one that includes both standard tools and libbpf-based tools?

**Comment by mauriciovasquezbernal** [comment]
> I am not sure to understand why we need to build those 3 flavours. Couldn't we just build the one that includes both standard tools and libbpf-based tools?

I want to build three different to optimize the deployment in case the user chooses which mode to use when deploying Inspektor Gadget. But it's also true that we could just ship everything in a single image.

File: Dockerfile.ubuntu [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ARG OS_TAG=18.04
+ARG OS_TAG=20.04
```
Comment (alban) [inline Dockerfile.ubuntu:1]: I wonder if that change could go upstream too.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -15,13 +15,14 @@ jobs:
...
-          OS_RELEASE: 18.04
         - NAME: focal-release
           OS_RELEASE: 20.04
     steps:
 
-    - uses: actions/checkout@v1
+    - uses: actions/checkout@v2
+      with:
+          submodules: recursive
+          fetch-depth: 0
```
Comment (alban) [inline .github/workflows/publish.yml:25]: This can be done upstream

File: Dockerfile.core [hunk 3]
```diff
@@ -0,0 +1,46 @@
+ARG OS_TAG=20.04
+FROM ubuntu:${OS_TAG} as builder
+
+ARG OS_TAG
+ARG BUILD_TYPE=release
+ARG DEBIAN_FRONTEND=noninteractive
```
Comment (alban) [inline Dockerfile.core:6]: Is it an `ARG` on purpose?

```suggestion
ENV DEBIAN_FRONTEND=noninteractive
```

(Or just add it in the relevant `RUN` commands like we do in Inspektor Gadget.)

File: Dockerfile.core [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+RUN cd /root/bcc/libbpf-tools/ && LLVM_STRIP=llvm-strip-10 make
+
+FROM ubuntu:${OS_TAG}
+
+RUN apt-get -qq update && \
+    apt-get -y install libelf1 libfl2
+
+# TODO: find a way to avoid specifying all the names
+COPY --from=builder                     \
+    /root/bcc/libbpf-tools/biolatency   \
```
Comment (alban) [inline Dockerfile.core:23]: Doesn't this work?

```suggestion
COPY --from=builder                     \
    /root/bcc/libbpf-tools/*   \
```

File: libbpf-tools/bindsnoop.bpf.c [hunk 5]
```diff
@@ -78,6 +88,12 @@ static int probe_exit(struct pt_regs *ctx, short ver)
 	if (filter_by_port && !port)
 		goto cleanup;
 
+	task = (struct task_struct*)bpf_get_current_task();
+	mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);
+
+	if (filter_by_mnt_ns && !bpf_map_lookup_elem(&mount_ns_set, &mntns_id))
+		return 0;
```
Comment (alban) [inline libbpf-tools/bindsnoop.bpf.c:95]: Should this be guarded by `filter_by_mnt_ns`? 

```suggestion
	if (filter_by_mnt_ns) {
		task = (struct task_struct*)bpf_get_current_task();
		mntns_id = (u64) BPF_CORE_READ(task, nsproxy, mnt_ns, ns.inum);

		if (!bpf_map_lookup_elem(&mount_ns_set, &mntns_id))
			return 0;
	}
```

File: Dockerfile.core (hunk 3 above)
Comment (mauriciovasquezbernal) [reply Dockerfile.core:6]: It's a copy-pasta error. Anyway I'll remove this file.

File: Dockerfile.core (hunk 4 above)
Comment (mauriciovasquezbernal) [reply Dockerfile.core:23]: That would also copy all the source files. Actually I solved it by using `make install` to a specific folder on the builder image and then copying all the contents of that folder to the final image.

File: libbpf-tools/bindsnoop.bpf.c (hunk 5 above)
Comment (mauriciovasquezbernal) [reply libbpf-tools/bindsnoop.bpf.c:95]: Well that's another way to write this. I think my way works fine and it relies on the compiler being smart enough to remove the non needed parts. Do you think we should rewrite it according to your suggestion?

File: .github/workflows/publish.yml (hunk 2 above)
Comment (mauriciovasquezbernal) [reply .github/workflows/publish.yml]: It's not needed when building a single container image for all the tools.

File: Dockerfile.ubuntu (hunk 1 above)
Comment (mauriciovasquezbernal) [reply Dockerfile.ubuntu]: It's not needed anymore.

**Review by alban — APPROVED** [review]
LGTM!

**Pull request merged**